sqlx-core/src/postgres/types/chrono/datetime.rs

```diff
@@ -101,21 +104,52 @@ impl<Tz: TimeZone> Encode<'_, Postgres> for DateTime<Tz> {
 
 impl<'r> Decode<'r, Postgres> for DateTime<Local> {
     fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
-        let naive = <NaiveDateTime as Decode<Postgres>>::decode(value)?;
-        Ok(Local.from_utc_datetime(&naive))
+        Ok(<DateTime<Utc> as Decode<Postgres>>::decode(value)?.with_timezone(&Local))
     }
 }
 
 impl<'r> Decode<'r, Postgres> for DateTime<Utc> {
     fn decode(value: PgValueRef<'r>) -> Result<Self, BoxDynError> {
-        let naive = <NaiveDateTime as Decode<Postgres>>::decode(value)?;
-        Ok(Utc.from_utc_datetime(&naive))
+        Ok(match value.format() {
+            PgValueFormat::Binary => {
+                // TIMESTAMP is encoded as the microseconds since the epoch
+                let epoch = Utc.ymd(2000, 1, 1).and_hms(0, 0, 0);
+                let us = Decode::<Postgres>::decode(value)?;
+                epoch + Duration::microseconds(us)
+            }
+
+            PgValueFormat::Text => {
+                let s = value.as_str()?;
+                DateTime::parse_from_str(
+                    s,
+                    if is_timestamptz_output(s) {
+                        // Note that because the output format differs from RFC3339,
+                        // `DateTime::parse_from_rfc3339` is not available
+                        "%Y-%m-%d %H:%M:%S%.f%#z"
+                    } else {
+                        "%Y-%m-%d %H:%M:%S%.f"
```

Comment:
`DateTime<Tz>` should reject a text format that doesn't contain a timezone offset, as that is a `timestamp without time zone` which cannot be placed on an actual timeline and is a symbolic datetime only;  the error could recommend `NaiveDateTime` as the more appropriate type.